server/apps/subscriptions.go

```diff
@@ -4,31 +4,149 @@
 	return &subscriptions{
+		mm:           mm,
 		configurator: configurator,
 	}
 }
 
-func (subs *subscriptions) GetSubscriptionsForChannel(subj SubscriptionSubject, channelID string) ([]*Subscription, error) {
-	return []*Subscription{
-		{
-			AppID:     "Hello",
-			Subject:   SubjectUserJoinedChannel,
-			ChannelID: "",
-		},
-	}, nil
+// GetSubscriptionsForChannelOrTeam returns subscriptions for a given subject and
+// channelID or teamID from the store
+func (subs *subscriptions) GetSubscriptionsForChannelOrTeam(subj SubscriptionSubject, channelOrTeamID string) ([]*Subscription, error) {
+	key := GetSubsKVkey(subj, channelOrTeamID)
+	var savedSubs []*Subscription
+	if err := subs.mm.KV.Get(key, &savedSubs); err != nil {
+		return nil, errors.Wrap(err, "failed to get saved subscriptions")
+	}
+
+	return savedSubs, nil
+}
+
+func (subs *subscriptions) GetSubscriptionsForApp(app string, subj SubscriptionSubject, channelID string) ([]*Subscription, error) {
+	// if subj is nil, grab all subjects for the
+	return nil, nil
+}
+
+// StoreSubscription stores a subscription for a change notification
+// TODO move this to store package or file
+func (subs *subscriptions) StoreSubscription(subj SubscriptionSubject, sub Subscription, channelID string) error {
+	key := GetSubsKVkey(subj, channelID)
+
+	// get all subscriptions for the subject
+	var savedSubs []*Subscription
+	if err := subs.mm.KV.Get(key, &savedSubs); err != nil {
+		return errors.Wrap(err, "failed to get saved subscriptions")
+	}
+
+	// check if sub exists
+	var newSubs []*Subscription
+	foundSub := 0
+	for _, s := range savedSubs {
+		// modify the sub to the latest request
+		if s.SubscriptionID == sub.SubscriptionID {
+			foundSub++
+			newSubs = append(newSubs, &sub)
+			continue
+		}
+		newSubs = append(newSubs, s)
+	}
+	if foundSub == 0 {
+		newSubs = append(newSubs, &sub)
+	}
+
+	// sub exists. update and save updated subs
+	_, err := subs.mm.KV.Set(key, newSubs)
+	if err != nil {
+		return errors.Wrap(err, "failed to save subscriptions")
+	}
+	return nil
+}
+
+// DeleteSubscription deletes a subscription
+func (subs *subscriptions) DeleteSubscription(subj SubscriptionSubject, subID SubscriptionID, channelID string) error {
+	key := GetSubsKVkey(subj, channelID)
+
+	// get all subscriptions for the subject
+	var savedSubs []*Subscription
+	if err := subs.mm.KV.Get(key, &savedSubs); err != nil {
+		return errors.Wrap(err, "failed to get saved subscriptions")
+	}
+
+	// check if sub exists
+	var newSubs []*Subscription
+	for _, s := range savedSubs {
+		if s.SubscriptionID == subID {
+			continue
+		}
+		newSubs = append(newSubs, s)
+	}
+
+	// sub was deleted. update and save updated subs
+	// TODO check for following:
+	//   - don't need to save if sub was not deleted?
+	//   - if delete the last subscription for the channel, delete the key also
+	_, err := subs.mm.KV.Set(key, newSubs)
```

Comment:
Should we worry about the lack of atomicity of this? We might end up with a race condition if two subscriptions where to be deleted nearly at the same time.

Reply:
1/5 I specifically asked @jfrerich not to worry concurrency yet. I fail to imagine this to be the final storage implementation. We just need a POC that works functionally. Having said that, _simple_ improvements would be welcome.